src/NumericalAlgorithms/Spectral/Spectral.cpp

```diff
@@ -369,50 +369,72 @@ Matrix interpolation_matrix(const size_t num_points,
-    const Mesh<1>& mesh) noexcept {
-  ASSERT(mesh.basis(0) == Spectral::Basis::Legendre,
-         "We only support DG with a Legendre basis.");
-  ASSERT(mesh.quadrature(0) == Spectral::Quadrature::Gauss,
-         "We only compute the boundary interpolation for Gauss quadrature "
-         "since for Gauss-Lobatto you can just copy values off the volume.");
-  static const auto cache = make_static_cache<CacheRange<
-      1, Spectral::maximum_number_of_points<Spectral::Basis::Legendre>>>(
-      [](const size_t num_points) noexcept {
-        const Mesh<1> local_mesh{num_points, Spectral::Basis::Legendre,
-                                 Spectral::Quadrature::Gauss};
-        return std::pair<Matrix, Matrix>{interpolation_matrix(local_mesh, -1.0),
-                                         interpolation_matrix(local_mesh, 1.0)};
+    const size_t num_points) noexcept {
+  static_assert(BasisType == Spectral::Basis::Legendre);
+  static_assert(QuadratureType == Spectral::Quadrature::Gauss);
+  static const auto cache = make_static_cache<
+      CacheRange<Spectral::minimum_number_of_points<BasisType, QuadratureType>,
+                 Spectral::maximum_number_of_points<BasisType>>>(
+      [](const size_t local_num_points) noexcept {
+        return std::pair<Matrix, Matrix>{
+            interpolation_matrix<BasisType, QuadratureType>(local_num_points,
+                                                            -1.0),
+            interpolation_matrix<BasisType, QuadratureType>(local_num_points,
+                                                            1.0)};
       });
-  return cache(mesh.number_of_grid_points());
+  return cache(num_points);
 }
 
-const std::pair<DataVector, DataVector>& boundary_lifting_term(
+const std::pair<Matrix, Matrix>& boundary_interpolation_matrix(
     const Mesh<1>& mesh) noexcept {
   ASSERT(mesh.basis(0) == Spectral::Basis::Legendre,
          "We only support DG with a Legendre basis.");
   ASSERT(mesh.quadrature(0) == Spectral::Quadrature::Gauss,
          "We only compute the boundary interpolation for Gauss quadrature "
          "since for Gauss-Lobatto you can just copy values off the volume.");
-  static const auto cache = make_static_cache<CacheRange<
-      1, Spectral::maximum_number_of_points<Spectral::Basis::Legendre>>>(
-      [](const size_t num_points) noexcept {
-        const Mesh<1> local_mesh{num_points, Spectral::Basis::Legendre,
-                                 Spectral::Quadrature::Gauss};
-        const auto& matrices = boundary_interpolation_matrix(local_mesh);
-        std::pair<DataVector, DataVector> result{
-            DataVector{local_mesh.number_of_grid_points()},
-            DataVector{local_mesh.number_of_grid_points()}};
-        for (size_t i = 0; i < local_mesh.number_of_grid_points(); ++i) {
+  return boundary_interpolation_matrix<Spectral::Basis::Legendre,
+                                       Spectral::Quadrature::Gauss>(
+      mesh.extents(0));
+}
```

Comment:
I think you can replace this function with:

```cpp
SPECTRAL_QUANTITY_FOR_MESH(boundary_interpolation_matrix, std::pair<Matrix, Matrix>)
```

(best placed below line 520. Same for the `boundary_lifting_term` function.)

Reply:
That doesn't have the `ASSERT`s in it

Reply:
The `boundary_interpolation_matrix` overload that's templated on `Basis` and `Quadrature` has the asserts, so that should be enough right

Reply:
Those are static asserts, we need a runtime check